README
This is the README file for CPANPLUS::Dist::Build, a CPANPLUS plugin
to install modules that use 'Module::Build' as their installer

Please refer to 'perldoc CPANPLUS::Dist::Build' after installation 
for details.

#####################################################################

* Description

Params::Check

    CPANPLUS::Dist::Build is a distribution class for Module::Build
    related modules.
    Using this package, you can create, install and uninstall perl
    modules. It inherits from CPANPLUS::Dist.
    
    Normal users won't have to worry about the interface to this module,
    as it functions transparently as a plug-in to CPANPLUS and will 
    just Do The Right Thing when it's loaded.

#####################################################################

* Installation

CPANPLUS::Dist::Build follows the standard perl module install process

perl Makefile.PL
make
make test
make install

The module uses no C or XS parts, so no c-compiler is required.

######################################################################

AUTHOR
    This module by Jos Boumans <kane@cpan.org>.

COPYRIGHT
    This module is copyright (c) 2005 Jos Boumans <kane@cpan.org>. All
    rights reserved.

    This library is free software; you may redistribute and/or modify it
    under the same terms as Perl itself.


Changes
Revision history for Perl extension CPANPLUS::Dist::Build.

0.80 Sun Jun  5 13:43:55 BST 2016
  - Include test_requires when falling back to older mechanism
    for determining prereqs

0.78 Wed Jun 11 18:06:59 BST 2014
  - Shield tests from PERL_INSTALL_QUIET env var

0.76 Wed Jan  8 08:46:22 GMT 2014
  - Don't send test reports during prepare() phase

0.74 Tue Aug  6 20:53:04 BST 2013
  - Remove Module::Pluggable prereq

0.72 Sun Jul 21 16:21:40 BST 2013
  - Actually resolve the issue where Build.PL doesn't
    generate a Build script. Regression since 0.68

0.70 Fri Jan 25 15:58:27 GMT 2013
  - Add deprecation usage to warn if modules are loaded
    from corelib. CPANPLUS::Dist::Build will be leaving
    core with v5.20.0 and will still be available on
    CPAN.

0.68 Wed Dec  5 00:17:04 GMT 2012
  - Revert change to wrapper command-line

0.66 Tue Dec  4 23:05:53 GMT 2012
  - Resolve an issue where Build.PL doesn't generate a
    Build script

0.64 Sat Oct 20 23:21:36 BST 2012
  - Update Module::Install, no functional changes.

0.62 Mon Feb 13 22:15:57 GMT 2012
  - Test reporting enhancements

0.60 Thu Oct 27 20:54:22 BST 2011
  - Eliminate the need to use 'perlwrapper' at all by using
    an equivalent one-liner.

0.58 Sun Sep  4 20:42:48 BST 2011
  - Added local::lib type behaviour detection to disable use of 'sudo'

0.56 Tue Apr 19 14:51:07 BST 2011
  - Display failing tests only once to terminal.

0.54 Mon Feb 28 11:52:04 GMT 2011
  - Only skip 'perl' as a prereq when CPANPLUS version is
    less than 0.9102

0.52 Fri Jan  7 22:23:03 GMT 2011
  - Apply blead patch from Peter Acklam

0.50 Thu Oct 28 23:02:19 BST 2010
  - We clear the relevant env vars that effect Module::Build
    during the 'install' test. local::lib and cpanm set these.

0.48 Sun Sep  5 15:14:17 BST 2010
  - Explicitly set the location of cpanp-run-perl when under PERL_CORE
    in t/inc/conf.pl
  - Make sure we find cpanp-run-perl in ../../utils when PERL_CORE
  - Correct spelling in POD, pointed out by H.Merijn Brand

0.46 Sat Jan 23 07:52:19 GMT 2010
  - actually make the prereq resolving fallback to _build/ querying
    if the 'prereq_data' action fails. *facepalm*

0.44 Wed Nov 18 12:37:03 GMT 2009
  - Make the prereq resolving fallback to _build/ querying 
    if the 'prereq_data' action fails. Spotted by Johan Vromans.

0.41_01 Thu Nov  5 14:07:48 GMT 2009
  - Added MYMETA support

0.40 Mon Sep  7 14:15:03 BST 2009
  - Change the location of the CPANPLUS bin directory when
    running under core (bingos)

0.38 Mon Aug 24 13:59:11 BST 2009
  - Cosmetic change to the test output message so it
    matches CPANPLUS::Dist::MM (bingos)

0.36 Sat Jul  4 14:51:42 BST 2009
  - Make install() honour the buildflags directive, 
    absence spotted by Matt Trout

0.35_03 Tue Jun 30 22:48:57 BST 2009
  - Switch to IPC::Cmd->can_capture_buffer wherever we 
    need to rely on what is captured ( VMS and MSWin32 
    currently ).
  - Make sure to use perlwrapper for Build.PL on VMS

0.35_02 Sat Jun 27 09:35:55 BST 2009
  - Don't use the perlwrapper on VMS, (Craig Berry)
  - Skip env test on VMS

0.35_01 Wed Jun 24 09:57:28 BST 2009
  - Fix to _find_prereqs() for VMS when running prereq_data
    Build target, reported by Craig Berry.

0.34 Sat Jun 13 09:50:24 BST 2009
  - Fixed the 02packages.details.txt.gz file in our fake CPAN
      CPANPLUS-0.87_01 now checks this file for sanity. Ours
      was not sane >:) Spotted by the ever vigilant Vincent Pit.

0.32 Mon Jun  1 21:43:52 BST 2009
  - Removed generation of .output files
  - Fixed the setting of PERL5_CPANPLUS_IS_EXECUTING env var in 
    prepare. Spotted by Matt Trout
  - Added 'perlwrapper' to all perl calls. Spotted by Matt Trout.

0.30 Mon May 18 16:23:14 BST 2009
  - [RT #46176] Skip XS tests when perl compiled without dynamic
    loading. Reported by Andy Dougherty

0.28 Fri May  8 10:58:42 BST 2009
  - Add the CPANPLUS bin/ dir to $ENV{PATH} if we are running
    under PERL_CORE. This fixes warnings during 'make test' in
    the core. No need for other users to upgrade.

0.26 Thu May  7 16:40:26 BST 2009
  - Integrated the t/inc/conf.pl and dummy-CPAN/ structure from
    CPANPLUS into the testsuite. This resolves the issue where 
    CPAN indexes were fetched and built during the tests. 
    Spotted by Steve Hay, fixed with help from Jos Boumans.

0.24 Mon Apr 27 16:55:37 BST 2009
  - Removed CPANPLUS::inc from the code as CPANPLUS-0.86 does not 
    include this now.

0.22 Tue Apr 14 11:01:12 BST 2009
  - Fix the installdirs in Makefile.PL

0.20 Tue Apr  7 17:01:47 BST 2009

  - Made MI::Respository only necessary on the author-side
    we should be installable on 5.6.* now. No other changes
    so no need to update in core.
    
0.18 Mon Mar 30 10:03:21 BST 2009

  - Amendments for the new save_state in dev CPANPLUS.
  - Added M::I::Repository functionality to Makefile.PL

0.16 Thu Mar 19 22:24:06 GMT 2009

  - Applied patch from Vincent Pit [RT #44426] 
    'buildflags of the form "--option value" aren't honoured'

0.14 Tue Mar 10 22:56:24 GMT 2009

  - Bump versions for a stable release

0.13_02 Mon Mar  9 13:36:15 GMT 2009

  - Resolved some issues on MSWin32 due to capturing of STDOUT/STDERR

0.13_01 Sun Mar  8 13:32:01 GMT 2009

  - Ported the 'configure_requires' functionality from CP::D::MM 

0.12 Mon Mar  2 13:33:47 GMT 2009

  - Fixed the version number in ::Constants

0.10 Mon Mar  2 07:36:41 GMT 2009

  - Added 'installdirs' directive to Makefile.PL to resolve
    [rt #43739] (bingos)
  - Only a version change no need to update core.

0.08 Sat Feb 28 07:53:52 GMT 2009

  - bumped Module::Build prerequisite to 0.32 (dagolden)
  - refactored documentation to reflect changes to code (bingos)
  - removed all refs to Module::Build object (bingos)
  - added LICENSE file (bingos)
  - added documentation to ::Constants (bingos)
  - added POD and POD coverage tests (bingos)

0.06_05 Wed Feb 25 22:24:50 GMT 2009


t/inc/conf.pl
### On VMS, the ENV is not reset after the program terminates.
### So reset it here explicitly
my ($old_env_path, $old_env_perl5lib);
BEGIN {
    use FindBin; 
    use File::Spec;
    
    ### paths to our own 'lib' and 'inc' dirs
    ### include them, relative from t/
    my @paths   = map { "$FindBin::Bin/$_" } qw[../lib inc];

    ### absolute'ify the paths in @INC;
    my @rel2abs = map { File::Spec->rel2abs( $_ ) }
                    grep { not File::Spec->file_name_is_absolute( $_ ) } @INC;
    
    ### use require to make devel::cover happy
    require lib;
    for ( @paths, @rel2abs ) { 
        my $l = 'lib'; 
        $l->import( $_ ) 
    }

    use Config;

    ### and add them to the environment, so shellouts get them
    $old_env_perl5lib = $ENV{'PERL5LIB'};
    $ENV{'PERL5LIB'}  = join $Config{'path_sep'}, 
                        grep { defined } $ENV{'PERL5LIB'}, @paths, @rel2abs;
    
    ### add CPANPLUS' bin dir to the front of $ENV{PATH}, so that cpanp-run-perl
    ### and friends get picked up, only under PERL_CORE though.
    if ( $ENV{PERL_CORE} ) {
       $old_env_path = $ENV{PATH};
       $ENV{'PATH'}  = join $Config{'path_sep'}, 
                    grep { defined } "$FindBin::Bin/../../../bin", $ENV{'PATH'};

    }

    ### Fix up the path to perl, as we're about to chdir
    ### but only under perlcore, or if the path contains delimiters,
    ### meaning it's relative, but not looked up in your $PATH
    $^X = File::Spec->rel2abs( $^X ) 
        if $ENV{PERL_CORE} or ( $^X =~ m|[/\\]| );

    ### chdir to our own test dir, so we know all files are relative 
    ### to this point, no matter whether run from perlcore tests or
    ### regular CPAN installs
    chdir "$FindBin::Bin" if -d "$FindBin::Bin"
}

BEGIN {
    use IPC::Cmd;
   
    ### Win32 has issues with redirecting FD's properly in IPC::Run:
    ### Can't redirect fd #4 on Win32 at IPC/Run.pm line 2801
    $IPC::Cmd::USE_IPC_RUN = 0 if $^O eq 'MSWin32';
    $IPC::Cmd::USE_IPC_RUN = 0 if $^O eq 'MSWin32';
}

### Use a $^O comparison, as depending on module at this time
### may cause weird errors/warnings
END {
    if ($^O eq 'VMS') {
        ### VMS environment variables modified by this test need to be put back
        ### path is "magic" on VMS, we can not tell if it really existed before
        ### this was run, because VMS will magically pretend that a PATH
        ### environment variable exists set to the current working directory
        $ENV{PATH} = $old_env_path;

        if (defined $old_env_perl5lib) {
            $ENV{PERL5LIB} = $old_env_perl5lib;
        } else {
            delete $ENV{PERL5LIB};
        }
    }
}

use strict;
use CPANPLUS::Configure;
use CPANPLUS::Error ();

use File::Path      qw[rmtree];
use FileHandle;
use File::Basename  qw[basename];

{   ### Force the ignoring of .po files for L::M::S
    $INC{'Locale::Maketext::Lexicon.pm'} = __FILE__;
    $Locale::Maketext::Lexicon::VERSION = 0;
}

my $Env = 'PERL5_CPANPLUS_TEST_VERBOSE';

# prereq has to be in our package file && core!
use constant TEST_CONF_PREREQ           => 'Cwd';   
use constant TEST_CONF_MODULE           => 'Foo::Bar::EU::NOXS';
use constant TEST_CONF_MODULE_SUB       => 'Foo::Bar::EU::NOXS::Sub';
use constant TEST_CONF_AUTHOR           => 'EUNOXS';
use constant TEST_CONF_INST_MODULE      => 'Foo::Bar';
use constant TEST_CONF_INVALID_MODULE   => 'fnurk';
use constant TEST_CONF_MIRROR_DIR       => 'dummy-localmirror';
use constant TEST_CONF_CPAN_DIR         => 'dummy-CPAN';
use constant TEST_CONF_CPANPLUS_DIR     => 'dummy-cpanplus';
use constant TEST_CONF_INSTALL_DIR      => File::Spec->rel2abs(
                                                File::Spec->catdir(      
                                                    TEST_CONF_CPANPLUS_DIR,
                                                    'install'
                                                )
                                            );       

### we might need this Some Day when we're installing into
### our own sandbox. see t/20.t for details
# use constant TEST_INSTALL_DIR       => do {
#     my $dir = File::Spec->rel2abs( 'dummy-perl' );
# 
#     ### clean up paths if we are on win32    
#     ### dirs with spaces will be.. bad :(
#     $^O eq 'MSWin32'
#         ? Win32::GetShortPathName( $dir )
#         : $dir;
# };        

# use constant TEST_INSTALL_DIR_LIB 
#     => File::Spec->catdir( TEST_INSTALL_DIR, 'lib' );
# use constant TEST_INSTALL_DIR_BIN 
#     => File::Spec->catdir( TEST_INSTALL_DIR, 'bin' );
# use constant TEST_INSTALL_DIR_MAN1 
#     => File::Spec->catdir( TEST_INSTALL_DIR, 'man', 'man1' );
# use constant TEST_INSTALL_DIR_MAN3
#     => File::Spec->catdir( TEST_INSTALL_DIR, 'man', 'man3' );
# use constant TEST_INSTALL_DIR_ARCH
#     => File::Spec->catdir( TEST_INSTALL_DIR, 'arch' );
# 
# use constant TEST_INSTALL_EU_MM_FLAGS =>
#     ' INSTALLDIRS=site' .
#     ' INSTALLSITELIB='     . TEST_INSTALL_DIR_LIB .
#     ' INSTALLSITEARCH='    . TEST_INSTALL_DIR_ARCH .    # .packlist
#     ' INSTALLARCHLIB='     . TEST_INSTALL_DIR_ARCH .    # perllocal.pod
#     ' INSTALLSITEBIN='     . TEST_INSTALL_DIR_BIN .
#     ' INSTALLSCRIPT='      . TEST_INSTALL_DIR_BIN .
#     ' INSTALLSITEMAN1DIR=' . TEST_INSTALL_DIR_MAN1 .
#     ' INSTALLSITEMAN3DIR=' . TEST_INSTALL_DIR_MAN3;


sub dummy_cpan_dir {
    ### VMS needs this in directory format for rel2abs
    my $test_dir = $^O eq 'VMS'
                    ? File::Spec->catdir(TEST_CONF_CPAN_DIR)
                    : TEST_CONF_CPAN_DIR;

    ### Convert to an absolute file specification
    my $abs_test_dir = File::Spec->rel2abs($test_dir);
    
    ### According to John M: the hosts path needs to be in UNIX format.  
    ### File::Spec::Unix->rel2abs does not work at all on VMS
    $abs_test_dir    = VMS::Filespec::unixify( $abs_test_dir ) if $^O eq 'VMS';

    return $abs_test_dir;
}

sub gimme_conf { 

    ### don't load any other configs than the heuristic one
    ### during tests. They might hold broken/incorrect data
    ### for our test suite. Bug [perl #43629] showed this.
    my $conf = CPANPLUS::Configure->new( load_configs => 0 );

    my $dummy_cpan = dummy_cpan_dir();
    
    $conf->set_conf( hosts  => [ { 
                        path        => $dummy_cpan,
                        scheme      => 'file',
                    } ],      
    );
    $conf->set_conf( base       => File::Spec->rel2abs(TEST_CONF_CPANPLUS_DIR));
    $conf->set_conf( dist_type  => '' );
    $conf->set_conf( signature  => 0 );
    $conf->set_conf( verbose    => 1 ) if $ENV{ $Env };
    
    ### never use a pager in the test suite
    $conf->set_program( pager   => '' );

    ### dmq tells us that we should run with /nologo
    ### if using nmake, as it's very noisy otherwise.
    {   my $make = $conf->get_program('make');
        if( $make and basename($make) =~ /^nmake/i ) {
            $conf->set_conf( makeflags => '/nologo' );
        }
    }

    $conf->set_conf( source_engine =>  $ENV{CPANPLUS_SOURCE_ENGINE} )
        if $ENV{CPANPLUS_SOURCE_ENGINE};
    
    _clean_test_dir( [
        $conf->get_conf('base'),     
        TEST_CONF_MIRROR_DIR,
#         TEST_INSTALL_DIR_LIB,
#         TEST_INSTALL_DIR_BIN,
#         TEST_INSTALL_DIR_MAN1, 
#         TEST_INSTALL_DIR_MAN3,
    ], (  $ENV{PERL_CORE} ? 0 : 1 ) );
        
    return $conf;
};

# placeholder

### clean these files if we're under perl core
END { 
    if ( $ENV{PERL_CORE} ) {
        close output_handle(); 1 while unlink output_file();

        _clean_test_dir( [
            gimme_conf->get_conf('base'),   
            TEST_CONF_MIRROR_DIR,
    #         TEST_INSTALL_DIR_LIB,
    #         TEST_INSTALL_DIR_BIN,
    #         TEST_INSTALL_DIR_MAN1, 
    #         TEST_INSTALL_DIR_MAN3,
        ], 0 ); # DO NOT be verbose under perl core -- makes tests fail
    }
}

### whenever we start a new script, we want to clean out our
### old files from the test '.cpanplus' dir..
sub _clean_test_dir {
    my $dirs    = shift || [];
    my $verbose = shift || 0;

    for my $dir ( @$dirs ) {

        ### no point if it doesn't exist;
        next unless -d $dir;

        my $dh;
        opendir $dh, $dir or die "Could not open basedir '$dir': $!";
        while( my $file = readdir $dh ) { 
            next if $file =~ /^\./;  # skip dot files
            
            my $path = File::Spec->catfile( $dir, $file );
            
            ### directory, rmtree it
            if( -d $path ) {

                ### John Malmberg reports yet another VMS issue:
                ### A directory name on VMS in VMS format ends with .dir 
                ### when it is referenced as a file.
                ### In UNIX format traditionally PERL on VMS does not remove the
                ### '.dir', however the VMS C library conversion routines do
                ### remove the '.dir' and the VMS C library routines can not 
                ### handle the '.dir' being present on UNIX format filenames.
                ### So code doing the fixup has on VMS has to be able to handle 
                ### both UNIX format names and VMS format names. 
                
                ### XXX See http://www.xray.mpe.mpg.de/
                ### mailing-lists/perl5-porters/2007-10/msg00064.html
                ### for details -- the below regex could use some touchups
                ### according to John. M.            
                $file =~ s/\.dir$//i if $^O eq 'VMS';
                
                my $dirpath = File::Spec->catdir( $dir, $file );

                print "# Deleting directory '$dirpath'\n" if $verbose;
                eval { rmtree( $dirpath ) };
                warn "Could not delete '$dirpath' while cleaning up '$dir'" 
                    if $@;
           
            ### regular file
            } else {
                print "# Deleting file '$path'\n" if $verbose;
                1 while unlink $path;
            }            
        }       
    
        close $dh;
    }
    
    return 1;
}
1;


Makefile.PL
use strict;
use inc::Module::Install;
name 'CPANPLUS-Dist-Build';
author 'Jos Boumans <kane[at]cpan.org>, Ken Williams <kwilliams@cpan.org>';
version_from 'lib/CPANPLUS/Dist/Build.pm';
abstract 'CPANPLUS plugin to install packages that use Build.PL';
perl_version '5.006';
license 'perl';
auto_license holder => 'Jos Boumans, Ken Williams, Chris Williams and David Golden';
build_requires 'Test::More'          => '0.47';
requires 'Test::Harness'             => '3.16';
requires 'CPANPLUS'                  => '0.84';
requires 'Locale::Maketext::Simple'  => 0;
requires 'Params::Check'             => '0.26';
requires 'IPC::Cmd'                  => '0.42';
requires 'Module::Load'              => '0.16';
requires 'Module::Load::Conditional' => '0.30';
requires 'Module::Build'             => '0.32';
requires 'File::Spec'                => 0;
requires 'ExtUtils::Install'         => '1.42';
requires 'if'                        => 0;
clean_files 't/dummy-cpanplus t/dummy-perl/lib/perl5';
# overwrite the CP::D::B that shipped in core
if ( $] >= 5.009005 && $] < 5.012 ) {
  install_as_core;
}
else {
  install_as_site;
}
eval "githubmeta";
auto_provides;
WriteAll();
